probe/ebpf/flow.c

```diff
@@ -397,47 +470,95 @@ int bpf_flow_table(struct __sk_buff *skb)
-
-	prev = bpf_map_lookup_element(&flow_table, &flow.key);
-	if (prev) {
-		update_metrics(skb, prev, tm,
-					   flow.link_layer._hash_src == prev->link_layer._hash_src);
-		__sync_fetch_and_add(&prev->last, tm - prev->last);
-
-		if (prev->layers & flow.layers & TRANSPORT_LAYER > 0) {
-			if (prev->transport_layer.port_src == flow.transport_layer.port_src) {
-				if (prev->transport_layer.ab_syn == 0 && flow.transport_layer.ab_syn != 0) {
-					__sync_fetch_and_add(&prev->transport_layer.ab_syn, flow.transport_layer.ab_syn);
-				}
-				if (prev->transport_layer.ab_fin == 0 && flow.transport_layer.ab_fin != 0) {
-					__sync_fetch_and_add(&prev->transport_layer.ab_fin, flow.transport_layer.ab_fin);
-				}
-				if (prev->transport_layer.ab_rst == 0 && flow.transport_layer.ab_rst != 0) {
-					__sync_fetch_and_add(&prev->transport_layer.ab_rst, flow.transport_layer.ab_rst);
-				}
-			}
-			else {
-				if (prev->transport_layer.ba_syn == 0 && flow.transport_layer.ab_syn != 0) {
-					__sync_fetch_and_add(&prev->transport_layer.ba_syn, flow.transport_layer.ab_syn);
-				}
-				if (prev->transport_layer.ba_fin == 0 && flow.transport_layer.ab_fin != 0) {
-					__sync_fetch_and_add(&prev->transport_layer.ba_fin, flow.transport_layer.ab_fin);
-				}
-				if (prev->transport_layer.ba_rst == 0 && flow.transport_layer.ab_rst != 0) {
-					__sync_fetch_and_add(&prev->transport_layer.ba_rst, flow.transport_layer.ab_rst);
-				}
-			}
-		}
-	} else {
-		update_metrics(skb, &flow, tm, 1);
-		__sync_fetch_and_add(&flow.start, tm);
-		__sync_fetch_and_add(&flow.last, tm);
+	__u32 k32 = 0;
+	struct l2 *l2 = bpf_map_lookup_element(&l2_table, &k32);
+	if (l2 == NULL) {
+		bpf_printk("=== l2 not found bpf_flow_table ===\n");
+		return 0;
+	}
+	memset(l2, 0, sizeof(struct l2));
+	l2->last = tm;
+
+	l2->link_layer.id = fill_l2(skb, l2);
+
+ 	bpf_tail_call(skb, &jmp_map, JMP_NETWORK_LAYER);
+	bpf_printk("=== no tail call l2 key %x %x ====\n", l2->key, l2->link_layer.id);
+	return 0;
+}
+
+SOCKET(network_layer)
+int network_layer(struct __sk_buff *skb)
+{
+	__u32 k32 = 0;
+	struct l2 *l2 = bpf_map_lookup_element(&l2_table, &k32);
+	if (l2 == NULL) {
+		bpf_printk("=== l2 not found network_layer ===\n");
+		return 0;
+	}
 
-		bpf_map_update_element(&flow_table, &flow.key, &flow, BPF_ANY);
+	struct flow *new, *flow;
+	new = bpf_map_lookup_element(&flow_nostack, &k32);
+	if (new == NULL) {
+		bpf_printk("=== no free struct flow\n");
+		return 0;
+	}
+	memset(new, 0, sizeof(struct flow));
+	
+	new->key = l2->key;
+	new->layers_path = l2->layers_path;
+	new->link_layer = l2->link_layer;
+	new->last = l2->last;
+
+	__u16 protocol = (__u16)l2->link_layer.id;
+	if (protocol != 0) {
+		new->layers_info |= LINK_LAYER_INFO;
+	}
+
+	/* L3/L4/... */
+	fill_network(skb, protocol, l2->next_layer_offset, skb->len - l2->next_layer_offset, new);
+
+	flow = bpf_map_lookup_element(&flow_table, &new->key);
+	if (flow == NULL) {
+		/* New flow */
+		new->ifindex = skb->ifindex;
```

Comment:
not related to this patch but doesn't seem to be used

Reply:
yes it's not used and it's always accessible via skb->ifindex for debug purpose .. removing it

Reply:
for GRE L2 it's another game (or any other L2 tunnel layer), it's a little bit more tricky and I think will be address by a tail_call of the inner packet